miniml/compiler/compile.scm

```diff
@@ -1729,7 +1729,7 @@
 
 (define (env-open env menv)
   (let* ((open-bindings (lambda (e me)
-           (vhash-fold-right (match-lambda*
+           (vhash-fold (match-lambda*
```

Comment:
Thinking about this more, it may be that both directions are equivalent as we consistently use `replace` so keys in the vhash should be unique, and we never run the fold action on the same key twice. However, if we did not have unicity, I suspect that `fold-right` would result in shadowing in the unexpected order (or I am missing something), which makes the explicit effort of using `fold-right` confusing to me.

Reply:
I think it should be `fold-right`: if there are multiple bindings in the original environment, then with `fold-right` they will be iterated in reverse order (with the "real" one last), so that the "real" one will be the last to replace the previous ones.
After some additional thinking, it looks like using `vhash-replace` is not a good idea due to the following cases (which should not happen often):
```ocaml
module X = struct let x = 0 end
module Y = struct let x = 1 open X end
let () = print_int Y.x
```

Reply:
Ah, the "something" I was missing is that the newer declarations are at the beginning of the table, not the end -- so indeed I agree that `fold-right` is the right choice here. I will amend the PR, thanks!

In your example, the "bug" is that `open X` replaces the `Export` binding of `x` with a `Local` one, so no `x` will remain at all in the binding. If we used `vhash-cons` in this scenario, we would keep all bindings in the hashtable, so filtering out the local bindings would keep the export binding(s) for `x`.

Reply:
I amended the code to keep the `fold-right` and add a comment. I did not try fixing the local-shadowing-export bug yet, I propose to do this after the PR is merged (to keep this PR a refactoring only).

Reply:
I agree to wait to fix this bug, let's merge this PR first.